prisma-fmt/src/references.rs

```diff
@@ -0,0 +1,277 @@
+use log::*;
+use lsp_types::{Location, ReferenceParams, Url};
+use psl::{
+    diagnostics::FileId,
+    error_tolerant_parse_configuration,
+    parser_database::ParserDatabase,
+    schema_ast::ast::{
+        AttributePosition, CompositeTypePosition, EnumPosition, Field, FieldId, FieldPosition, FieldType, Identifier,
+        ModelId, ModelPosition, SchemaPosition, SourcePosition, Top, WithIdentifier, WithName,
+    },
+    Diagnostics, SourceFile,
+};
+
+use crate::{offsets::position_to_offset, span_to_range, LSPContext};
+
+pub(super) type ReferencesContext<'a> = LSPContext<'a, ReferenceParams>;
+
+pub(crate) fn empty_references() -> Vec<Location> {
+    Vec::new()
+}
+
+fn empty_identifiers<'ast>() -> Vec<&'ast Identifier> {
+    Vec::new()
+}
+
+pub(crate) fn references(schema_files: Vec<(String, SourceFile)>, params: ReferenceParams) -> Vec<Location> {
+    info!("Finding references");
+
+    let (_, config, _) = error_tolerant_parse_configuration(&schema_files);
+
+    let db = {
+        let mut diag = Diagnostics::new();
+        ParserDatabase::new(&schema_files, &mut diag)
+    };
+
+    let Some(initiating_file_id) = db.file_id(params.text_document_position.text_document.uri.as_str()) else {
+        warn!("Initating file name is not found in the schema");
+        return empty_references();
+    };
+
+    let initiating_doc = db.source(initiating_file_id);
+
+    let position = if let Some(pos) = position_to_offset(&params.text_document_position.position, initiating_doc) {
+        pos
+    } else {
+        warn!("Received a position outside of the document boundaries in ReferenceParams");
+        return empty_references();
+    };
+
+    let target_position = db.ast(initiating_file_id).find_at_position(position);
+
+    let ctx = ReferencesContext {
+        db: &db,
+        config: &config,
+        initiating_file_id,
+        params: &params,
+    };
+
+    reference_locations_for_target(ctx, target_position)
+}
+
+fn reference_locations_for_target(ctx: ReferencesContext<'_>, target: SchemaPosition) -> Vec<Location> {
+    let identifiers: Vec<&Identifier> = match target {
+        // Blocks
+        SchemaPosition::Model(_, ModelPosition::Name(name))
+        | SchemaPosition::Enum(_, EnumPosition::Name(name))
+        | SchemaPosition::CompositeType(_, CompositeTypePosition::Name(name)) => {
+            find_where_used_as_top_name(&ctx, name)
+                .into_iter()
+                .chain(find_where_used_as_type(&ctx, name))
+                .collect()
```

Comment:
Do you need to collect all of these vecs? My intuition is no. You're collecting a `Vec<Location>` below anyway and each case here could just return iterators of `&Identifier`

Reply:
in the meantime, `find_where_used_for_native_type` has been updated as that will most likely be the most intensive call when used